File: tools/debugging/parse_rsp_profile.py
#!/usr/bin/env python3

"""
This script parses RSP profiler output to generate a summary report and heatmap
of frame performance hot spots.

Usage:

1. Build the game with hardware debugging support and the RSP profiler enabled.

2. While debugging the game, press d-pad down on controller 3 to log timing
   information and save screenshots for the last rendered frame's display list.

3. Run this script with the debug log file passed via --debug-log-file. It will
   output a report of the cost of each captured display list command.

   The log file is also used to label dynamic model display lists. Optionally
   provide portal.map from the build directory using --symbol-map-file to label
   other display lists as well.

   Specify the saved screenshots' directory with --screenshot-dir to generate a
   heatmap highlighting pixel costs.

Caveats:

* Only commands from the top-level display list are present in the final output.
  The profiler does not recursively measure child performance.

* Dynamically loaded elements such as level geometry can share addresses with
  each other. Such display lists are labelled using a comma-separated list of
  all matching symbol names.

* The game builds some display lists at runtime, meaning they don't have symbol
  names. They are labelled using a +-separated list of all named child display
  lists (recursively). The top-level address is used when no names are found.

* Due to performance fluctuations and the way time is measured, some fast
  commands can be shown with negative run times.
"""

import argparse
import datetime
import itertools
import math
import os
import re
import struct

class SymbolMap:
    STATIC_SYMBOL_REGEX = re.compile(r"^\s+0x([a-f0-9]+)\s+(\w+)$")
    DYNAMIC_ASSET_RESET_REGEX = re.compile(r"^Reset dynamic assets$")
    DYNAMIC_ASSET_LOAD_REGEX = re.compile(r"^Loaded dynamic asset at 0x([a-f0-9]{8}): (\w+)$")

    def __init__(self, file_path):
        self._static_symbols = file_path and self._parse_symbol_file(file_path) or dict()
        self._dynamic_symbols = dict()

    @classmethod
    def _parse_static_symbol(cls, line):
        match = cls.STATIC_SYMBOL_REGEX.match(line)
        if not match:
            return None, None

        address, name = match.groups()
        return int(address, 16), name

    @classmethod
    def _parse_symbol_file(cls, file_path):
        address_to_symbol = dict()

        with open(file_path, "r") as f:
            for line in f:
                address, name = cls._parse_static_symbol(line)
                if address:
                    # Dynamic elements can share the same address
                    names = filter(None, (address_to_symbol.get(address), name))
                    address_to_symbol[address] = ",".join(names)

        return address_to_symbol

    @classmethod
    def _is_dynamic_asset_reset_line(cls, line):
        return cls.DYNAMIC_ASSET_RESET_REGEX.match(line)

    @classmethod
    def _parse_dynamic_asset_load(cls, line):
        match = cls.DYNAMIC_ASSET_LOAD_REGEX.match(line)
        if not match:
            return None, None

        address, name = match.groups()
        return int(address, 16), name

    def try_update(self, line):
        if (self._is_dynamic_asset_reset_line(line)):
            self._dynamic_symbols.clear()
            return True

        address, name = self._parse_dynamic_asset_load(line)
        if address:
            self._dynamic_symbols[address] = name
            return True

        return False

    def get_symbol_name(self, address):
        return self._static_symbols.get(address) or \
            self._dynamic_symbols.get(address)

class DisplayList:
    G_FILLRECT = 0xf6
    G_POPMTX   = 0xd8
    G_MTX      = 0xda
    G_MOVEWORD = 0xdb
    G_DL       = 0xde
    G_ENDDL    = 0xdf

    class Command:
        COMMAND_REGEX = re.compile(r"^dl 0x([a-f0-9]{2})([a-f0-9]{6})([a-f0-9]{8})$")

        def __init__(self, command, w0, w1):
            self.command = command
            self.w0 = w0
            self.w1 = w1
            self.children = []
            self.name = None

        @classmethod
        def parse(cls, line):
            match = cls.COMMAND_REGEX.match(line)
            if not match:
                return None

            command, w0, w1 = match.groups()
            return cls(
                int(command, 16),
                int(w0, 16),
                int(w1, 16)
            )

    def __init__(self, symbol_map):
        self._symbol_map = symbol_map
        self._stack = [None]  # Dummy element for root display list
        self._visited = set()
        self._child_names = dict()

    def _update(self, command):
        match command.command:
            case self.G_DL:
                current = self._stack[-1]
                if current:
                    current.children.append(command)
                self._stack.append(command)

            case self.G_ENDDL:
                if not self._stack:
                    raise ValueError("Malformed display list")

                dl = self._stack.pop()
                if dl and not (dl in self._visited):
                    name = self._symbol_map.get_symbol_name(dl.w1) or \
                        "+".join(filter(None, (c.name for c in dl.children)))

                    if name:
                        dl.name = name
                        self._child_names[dl.w1] = dl.name

                    self._visited.add(dl.w1)

    def try_update(self, line):
        command = self.Command.parse(line)
        if command:
            self._update(command)
            return True

        return False

    def get_command_name(self, command):
        match command.command:
            case self.G_FILLRECT:
                return "gsDPFillRectangle"
            case self.G_POPMTX:
                return "gsSPPopMatrix"
            case self.G_MTX:
                return "gsSPMatrix"
            case self.G_MOVEWORD:
                segment_num = (command.w0 // 4) & 0xf
                return f"gsSPSegment(0x{segment_num:x}, 0x{command.w1:08x})"
            case self.G_DL:
                address = self._child_names.get(command.w1) or \
                    f"0x{command.w1:08x}"
                return f"gsSPDisplayList({address})"
            case _:
                return f"unknown 0x{command.command:x} 0x{command.w0:08x}{command.w1:08x}"

class CommandInfo:
    SAMPLE_REGEX = re.compile(r"^(\d+)\/\d+ 0x([a-f0-9]{2})([a-f0-9]{6})([a-f0-9]{8}) (\d+\.?\d+) ms$")
    SCREENSHOT_NAME_REGEX = re.compile(r"^.+'(.+\.bin)'.+$")

    def __init__(self, command_index, command, w0, w1, start_time_ms):
        self.command_index = command_index
        self.command = command
        self.w0 = w0
        self.w1 = w1
        self.start_time_ms = start_time_ms
        self.sample_count = 1

        self.run_time_ms = 0
        self.screenshot = None
        self.diff_pixels = []

    def __iadd__(self, other):
        self.start_time_ms += other.start_time_ms
        self.sample_count += 1
        return self

    def try_update(self, line, screenshot_dir):
        match = self.SCREENSHOT_NAME_REGEX.match(line)
        if not match:
            return False

        screenshot_path = os.path.join(screenshot_dir, match[1])

        if os.path.exists(screenshot_path):
            with open(screenshot_path, "rb") as s:
                # Load up front for performance
                # This will stay in memory longer than needed, but it is simpler
                self.screenshot = list(itertools.batched(s.read(), 2, strict=True))

        return True

    def finish(self):
        self.start_time_ms /= self.sample_count

    def compute_diff(self, next):
        self.run_time_ms = next.start_time_ms - self.start_time_ms

        if self.screenshot and next.screenshot:
            self.diff_pixels = [
                i for i, (px1, px2) in enumerate(zip(self.screenshot, next.screenshot))
                if px1 != px2
            ]

    @classmethod
    def parse(cls, line):
        match = cls.SAMPLE_REGEX.match(line)
        if not match:
            return None

        command_index, command, w0, w1, start_time_ms = match.groups()
        return cls(
            int(command_index),
            int(command, 16),
            int(w0, 16),
            int(w1, 16),
            float(start_time_ms)
        )

class Heatmap:
    WIDTH = 320
    HEIGHT = 240

    BMP_HEADER_SIZE = 14
    BMP_DIB_HEADER_SIZE = 12
    BMP_BPP = 24

    def __init__(self):
        self._pixel_costs = dict()

    def update(self, pixel_indices, element_cost):
        if not pixel_indices or element_cost < 0:
            return False

        # Highlight elements with high per-pixel cost.
        #
        # Large expensive objects are more tolerable than small ones, as they
        # cover more of the screen. Small objects are more likely to appear
        # multiple times, or alongside other expensive objects.
        average_cost = element_cost / len(pixel_indices)
        for i in pixel_indices:
            self._pixel_costs[i] = self._pixel_costs.get(i, 0) + average_cost
        return True

    def write_bitmap(self, file_path):
        max_cost = max(self._pixel_costs.values())

        with open(file_path, "wb") as f:
            # Header and DIB header
            data_offset = self.BMP_HEADER_SIZE + self.BMP_DIB_HEADER_SIZE
            data_size = self.WIDTH * self.HEIGHT * (self.BMP_BPP // 8)
            f.write(struct.pack("<2cI4xI", b"\x42", b"\x4d", data_offset + data_size, data_offset))
            f.write(struct.pack("<I4H", self.BMP_DIB_HEADER_SIZE, self.WIDTH, self.HEIGHT, 1, self.BMP_BPP))

            # Data
            for row in range(self.HEIGHT):
                for col in range(self.WIDTH):
                    # Bitmaps store lines in reverse order
                    px_row = self.HEIGHT - row - 1
                    px_idx = (self.WIDTH * px_row) + col

                    value = math.floor((255 * self._pixel_costs.get(px_idx, 0) / max_cost) + 0.5)
                    f.write(struct.pack("3B", value, value, value))

class Profile:
    BEGIN_PROFILE_REGEX = re.compile(r"^Begin RSP profile$")
    END_PROFILE_REGEX = re.compile(r"^End RSP profile$")

    def __init__(self, symbol_map, screenshot_dir):
        self._name = datetime.datetime.now().strftime(f"profile-%y%m%d-%H%M%S%f")
        self._display_list = DisplayList(symbol_map)
        self._screenshot_dir = screenshot_dir
        self._command_info = []
        self._is_finished = False

    def _finish(self):
        for i in range(len(self._command_info)):
            current = self._command_info[i]
            current.finish()

            if i > 0:
                prev = self._command_info[i - 1]
                prev.compute_diff(current)

        # The last command is always a pipe sync we don't care about
        self._command_info.pop()
        self._is_finished = True

    @classmethod
    def is_profile_start_line(cls, line):
        return cls.BEGIN_PROFILE_REGEX.match(line)

    @classmethod
    def is_profile_end_line(cls, line):
        return cls.END_PROFILE_REGEX.match(line)

    def try_update(self, line):
        if (self.is_profile_end_line(line)):
            self._finish()
            return True

        info = CommandInfo.parse(line)
        if info:
            if info.command_index == len(self._command_info):
                self._command_info.append(info)
            elif info.command_index == len(self._command_info) - 1:
                self._command_info[-1] += info
            else:
                raise IndexError(f"Sample indices must be sequential")
            return True

        if self._command_info and self._command_info[-1].try_update(line, self._screenshot_dir):
            return True

        return self._display_list.try_update(line)

    def name(self):
        return self._name

    def is_finished(self):
        return self._is_finished

    def format_stats(self, sort=True):
        info = self._command_info[:]
        if sort:
            info.sort(
                key=lambda i: i.run_time_ms,
                reverse=True
            )

        column_padding = [20, 6, 0]
        pad = lambda e: f"{e[1]}".ljust(column_padding[e[0]])

        lines = [
            ["Run time (ms)", "Idx", "Command"],
            ["-------------", "---", "-------"],
            *[[
                f"{c.run_time_ms:.16f}",
                c.command_index,
                self._display_list.get_command_name(c)
            ] for c in info]
        ]

        return "\n".join(
            " ".join(pad(e) for e in enumerate(line))
            for line in lines
        )

    def build_heatmap(self):
        heatmap = Heatmap()

        num_updates = sum(
            1
            for info in self._command_info
            if heatmap.update(info.diff_pixels, info.run_time_ms)
        )

        return (num_updates > 0 and heatmap) or None

    def write(self, output_dir, sort=True):
        file_name_base = os.path.join(output_dir, f"{self._name}")

        with open(f"{file_name_base}.txt", "w") as f:
            f.write(self.format_stats(sort))

        heatmap = self.build_heatmap()
        if heatmap:
            heatmap.write_bitmap(f"{file_name_base}.bmp")

# Main

def get_args():
    parser = argparse.ArgumentParser(
        prog="parse_rsp_profile",
        description="Generates frame performance data from RSP profiler output"
    )
    parser.add_argument(
        "-d", "--debug-log-file",
        required=True,
        help="File containing debug console output"
    )
    parser.add_argument(
        "-m", "--symbol-map-file",
        help="Linker-generated file containing symbol names and addresses"
    )
    parser.add_argument(
        "-s", "--screenshot-dir",
        default=os.getcwd(),
        help="Directory containing screenshot files (defaults to current directory)"
    )
    parser.add_argument(
        "-o", "--output-dir",
        default=os.getcwd(),
        help="Directory to output results to (defaults to current directory)"
    )
    parser.add_argument(
        "-u", "--unsorted",
        action="store_true",
        help="Output the summary without sorting by command run time"
    )

    return parser.parse_args()

args = get_args()

symbol_map = SymbolMap(args.symbol_map_file)
current_profile = None

with open(args.debug_log_file, "r") as f:
    for line in f:
        if Profile.is_profile_start_line(line):
            current_profile = Profile(symbol_map, args.screenshot_dir)

        if current_profile:
            if current_profile.try_update(line):
                continue
            elif current_profile.is_finished():
                current_profile.write(args.output_dir, sort=not args.unsorted)
                print(f"Wrote {current_profile.name()} to {args.output_dir}")
                current_profile = None
                continue

        symbol_map.try_update(line)

File: tools/debugging/parse_rsp_profile.js
// This script parses RSP profiler output to generate a summary report and
// heatmap of frame performance hot spots.
//
// Usage:
// 1. Build the game with hardware debugging and the RSP profiler enabled.
//
// 2. While debugging the game, press d-pad down on controller 3 to log timing
//    information and save screenshots for the last frame's display list.
//
// 3. Run this script with the debug log file passed via --debug-log-file. It
//    will output a display list command cost report for each profiled frame.
//
//    The log file is also used to label dynamic model display lists. Optionally
//    provide portal.map from the build directory using --symbol-map-file to
//    label other display lists as well.
//
//    Specify the saved screenshots' directory with --screenshot-dir to generate
//    heatmaps highlighting pixel costs.
//
// Caveats:
// * Only commands from the top-level display list are included in cost reports.
//   The profiler does not recursively measure child performance.
//
// * Due to performance fluctuations and the way time is measured, some fast
//   commands can be shown with negative run times.
//
// * Dynamically loaded elements such as level geometry can share addresses with
//   each other. Such display lists are labelled using a comma-separated list of
//   all matching symbol names.
//
// * The game builds some display lists at runtime, meaning they have no symbol
//   names. They are labelled using a +-separated list of all named children
//   (recursively). The top-level address is used when no names are found.
//
// * Heatmap pixel values are computed by averaging each display list command's
//   cost across all pixels it affects, then summing the results.

const fs = require("fs");
const path = require("path");
const util = require("util");

const SCREEN_WD = 320;
const SCREEN_HT = 240;

class SymbolMap {
    static STATIC_SYMBOL_REGEX = /^0x([a-f0-9]+)\s+([a-zA-Z_]\w+)$/;
    static DYNAMIC_ASSET_RESET_REGEX = /^Reset dynamic assets$/
    static DYNAMIC_ASSET_LOAD_REGEX = /^Loaded dynamic asset at 0x([a-f0-9]{8}): (\w+)$/

    constructor(filePath) {
        this._staticSymbols = filePath
            ? SymbolMap._parseSymbolFile(filePath)
            : new Map();
        this._dynamicSymbols = new Map();
    }

    static _parseStaticSymbol(line) {
        const match = line.match(SymbolMap.STATIC_SYMBOL_REGEX);
        if (!match) {
            return null;
        }

        return {
            address: Number.parseInt(match[1], 16),
            name: match[2]
        };
    }

    static _parseSymbolFile(filePath) {
        const addressToSymbol = new Map();
        const lines = fs.readFileSync(filePath, "utf-8").split("\n");

        for (const line of lines) {
            const symbol = SymbolMap._parseStaticSymbol(line.trim());
            if (symbol) {
                // Dynamic elements can share the same address (multiple names)
                const names = [addressToSymbol.get(symbol.address), symbol.name];
                addressToSymbol.set(symbol.address, names.filter(Boolean).join(","))
            }
        }

        return addressToSymbol;
    }

    static _isDynamicAssetResetLine(line) {
        return SymbolMap.DYNAMIC_ASSET_RESET_REGEX.test(line);
    }

    static _parseDynamicAssetLoad(line) {
        const match = line.match(SymbolMap.DYNAMIC_ASSET_LOAD_REGEX);
        if (!match) {
            return null;
        }

        return {
            address: Number.parseInt(match[1], 16),
            name: match[2]
        }
    }

    tryUpdate(line) {
        if (SymbolMap._isDynamicAssetResetLine(line)) {
            this._dynamicSymbols.clear();
            return true;
        }

        const symbol = SymbolMap._parseDynamicAssetLoad(line);
        if (symbol) {
            this._dynamicSymbols.set(symbol.address, symbol.name);
            return true;
        }

        return false;
    }

    getSymbolName(address) {
        return this._staticSymbols.get(address) || this._dynamicSymbols.get(address);
    }
}

class DisplayList {
    static COMMAND_REGEX = /^dl 0x([a-f0-9]{2})([a-f0-9]{6})([a-f0-9]{8})$/;

    static G_FILLRECT = 0xf6;
    static G_POPMTX   = 0xd8;
    static G_MTX      = 0xda;
    static G_MOVEWORD = 0xdb;
    static G_DL       = 0xde;
    static G_ENDDL    = 0xdf;

    constructor(symbolMap) {
        this._symbolMap = symbolMap;
        this._stack = [null];  // Dummy element for root display list
        this._visited = new Set();
        this._childNames = new Map();
    }

    static _parseCommand(line) {
        const match = line.match(DisplayList.COMMAND_REGEX);
        if (!match) {
            return null;
        }

        return {
            command: Number.parseInt(match[1], 16),
            w0: Number.parseInt(match[2], 16),
            w1: Number.parseInt(match[3], 16),
            children: [],
            name: undefined
        };
    }

    _update(command) {
        switch (command.command) {
            case DisplayList.G_DL:
                const current = this._stack.at(-1);
                if (current) {
                    current.children.push(command);
                }
                this._stack.push(command);
                break;

            case DisplayList.G_ENDDL:
                if (this._stack.length === 0) {
                    throw new Error("Malformed display list");
                }

                const dl = this._stack.pop();
                if (!dl || this._visited.has(dl.w1)) {
                    break;
                }

                // Returning from display list. Try to name it using found info.
                const name = this._symbolMap.getSymbolName(dl.w1) ||
                    dl.children.map(c => c.name).filter(Boolean).join("+");

                if (name) {
                    dl.name = name;
                    this._childNames.set(dl.w1, dl.name);
                }

                this._visited.add(dl.w1);
                break;
        }
    }

    tryUpdate(line) {
        const command = DisplayList._parseCommand(line);
        if (command) {
            this._update(command);
            return true;
        }

        return false;
    }

    getCommandName(command) {
        const hexStr = (num, pad=0) => `${num.toString(16).padStart(pad, "0")}`;

        switch (command.command) {
            case DisplayList.G_FILLRECT:
                return `gsDPFillRectangle`;
            case DisplayList.G_POPMTX:
                return `gsSPPopMatrix`;
            case DisplayList.G_MTX:
                return `gsSPMatrix`;
            case DisplayList.G_MOVEWORD:
                const segmentNum = Math.trunc(command.w0 / 4) & 0xf;
                return `gsSPSegment(0x${hexStr(segmentNum)}, 0x${hexStr(command.w1, 8)})`;
            case DisplayList.G_DL:
                const address = this._childNames.get(command.w1) || `0x${hexStr(command.w1, 8)}`;
                return `gsSPDisplayList(${address})`;
            default:
                return `unknown 0x${hexStr(command.command)} 0x${hexStr(command.w0, 8)}${hexStr(command.w1, 8)}`;
        }
    }
}

class Heatmap {
    static BMP_HEADER_SIZE = 14;
    static BMP_DIB_HEADER_SIZE = 12;
    static BMP_BYTES_PER_PIXEL = 3;

    constructor() {
        this._pixelCosts = new Map();
    }

    update(pixelIndices, elementCost) {
        if (pixelIndices.length === 0 || elementCost <= 0) {
            return;
        }

        // Highlight elements with high per-pixel cost.
        //
        // Large expensive objects are more tolerable than small ones, as they
        // cover more of the screen. Small objects are more likely to appear
        // multiple times, or alongside other expensive objects.
        const averageCost = elementCost / pixelIndices.length;

        for (const i of pixelIndices) {
            const existing = this._pixelCosts.get(i) || 0;
            this._pixelCosts.set(i, existing + averageCost);
        }
    }

    isEmpty() {
        return this._pixelCosts.size === 0;
    }

    writeBitmap(filePath) {
        const maxCost = Math.max(...this._pixelCosts.values());

        const dataOffset = Heatmap.BMP_HEADER_SIZE + Heatmap.BMP_DIB_HEADER_SIZE;
        const dataSize = SCREEN_WD * SCREEN_HT * Heatmap.BMP_BYTES_PER_PIXEL;
        const buffer = Buffer.alloc(dataOffset + dataSize);

        // Header
        buffer.write("BM", 0);
        buffer.writeUInt32LE(buffer.length, 2);
        buffer.writeUInt32LE(0, 6);
        buffer.writeUInt32LE(dataOffset, 10);

        // DIB header
        buffer.writeUInt32LE(Heatmap.BMP_DIB_HEADER_SIZE, 14);
        buffer.writeUInt16LE(SCREEN_WD, 18);
        buffer.writeUInt16LE(SCREEN_HT, 20);
        buffer.writeUInt16LE(1, 22);
        buffer.writeUInt16LE(Heatmap.BMP_BYTES_PER_PIXEL * 8, 24);

        // Data
        for (let row = 0; row < SCREEN_HT; ++row) {
            for (let col = 0; col < SCREEN_WD; ++col) {
                // Bitmaps store lines in reverse order
                const inRow = SCREEN_HT - row - 1;
                const inIdx = (SCREEN_WD * inRow) + col;

                const cost = this._pixelCosts.get(inIdx) || 0;
                const value = Math.trunc((255 * cost / maxCost) + 0.5);

                const outIdx = dataOffset + ((SCREEN_WD * row) + col) * Heatmap.BMP_BYTES_PER_PIXEL;
                for (let i = 0; i < Heatmap.BMP_BYTES_PER_PIXEL; ++i) {
                    buffer.writeUInt8(value, outIdx + i);
                }
            }
        }

        fs.writeFileSync(filePath, buffer);
    }
}

class CommandInfo {
    static SAMPLE_REGEX = /^(\d+)\/\d+ 0x([a-f0-9]{2})([a-f0-9]{6})([a-f0-9]{8}) (\d+\.?\d+) ms$/;
    static SCREENSHOT_NAME_REGEX = /^.+'(.+\.bin)'.+$/;

    constructor(commandIndex, command, w0, w1, startTimeMs) {
        this.commandIndex = commandIndex;
        this.command = command;
        this.w0 = w0;
        this.w1 = w1;
        this.startTimeMs = startTimeMs;
        this.sampleCount = 1;

        this.runTimeMs = 0;
        this.screenshot = null;
        this.diffPixels = [];
    }

    static parse(line) {
        const match = line.match(CommandInfo.SAMPLE_REGEX);
        if (!match) {
            return null;
        }

        return new CommandInfo(
            Number(match[1]),
            Number.parseInt(match[2], 16),
            Number.parseInt(match[3], 16),
            Number.parseInt(match[4], 16),
            Number(match[5])
        );
    }

    tryUpdate(line, screenshotDir) {
        const match = line.match(CommandInfo.SCREENSHOT_NAME_REGEX);
        if (!match) {
            return false;
        }

        const screenshotPath = path.join(screenshotDir, match[1]);
        if (fs.existsSync(screenshotPath)) {
            // Load up front for performance
            // This will stay in memory longer than needed, but it is simpler
            this.screenshot = fs.readFileSync(screenshotPath);
        }

        return true;
    }

    accumulate(other) {
        this.startTimeMs += other.startTimeMs;
        this.sampleCount += other.sampleCount;
    }

    finish() {
        this.startTimeMs /= this.sampleCount;
    }

    computeDiff(next) {
        this.runTimeMs = next.startTimeMs - this.startTimeMs;

        if (this.screenshot && next.screenshot) {
            for (let pxIdx = 0; pxIdx < SCREEN_HT * SCREEN_WD; ++pxIdx) {
                const bufIdx = pxIdx * 2;

                if (this.screenshot[bufIdx + 0] !== next.screenshot[bufIdx + 0] ||
                    this.screenshot[bufIdx + 1] !== next.screenshot[bufIdx + 1]) {

                    this.diffPixels.push(pxIdx);
                }
            }
        }
    }
}

class Profile {
    static BEGIN_PROFILE_REGEX = /^Begin RSP profile$/;
    static END_PROFILE_REGEX = /^End RSP profile$/;

    constructor(symbolMap, screenshotDir) {
        const now = new Date();
        this.name = `profile-` +
            `${now.getFullYear() % 100}` +
            `${now.getMonth() + 1}`.padStart(2, "0") +
            `${now.getDate()}`.padStart(2, "0") + `-` +
            `${now.getHours()}`.padStart(2, "0") +
            `${now.getMinutes()}`.padStart(2, "0") +
            `${now.getSeconds()}`.padStart(2, "0") +
            `${now.getMilliseconds()}`.padStart(3, "0");
        this.isFinished = false;

        this._displayList = new DisplayList(symbolMap);
        this._screenshotDir = screenshotDir;
        this._commandInfo = [];
    }

    static isProfileStartLine(line) {
        return Profile.BEGIN_PROFILE_REGEX.test(line);
    }

    static isProfileEndLine(line) {
        return Profile.END_PROFILE_REGEX.test(line);
    }

    _finish() {
        for (let i = 0; i < this._commandInfo.length; ++i) {
            const current = this._commandInfo[i];
            current.finish();

            const prev = this._commandInfo[i - 1];
            prev?.computeDiff(current);
        }

        // The last command is always a pipe sync we don't care about
        this._commandInfo.pop();
        this.isFinished = true;
    }

    tryUpdate(line) {
        if (Profile.isProfileEndLine(line)) {
            this._finish();
            return true;
        }

        const info = CommandInfo.parse(line);
        if (info) {
            if (info.commandIndex === this._commandInfo.length) {
                this._commandInfo.push(info);
            } else if (info.commandIndex === this._commandInfo.length - 1) {
                this._commandInfo.at(-1).accumulate(info);
            } else {
                throw new Error("Sample indices must be sequential");
            }

            return true;
        }

        if (this._commandInfo.at(-1)?.tryUpdate(line, this._screenshotDir)) {
            return true;
        }

        return this._displayList.tryUpdate(line);
    }

    formatStats(sort=true) {
        const info = this._commandInfo.slice();
        if (sort) {
            info.sort((a, b) => b.runTimeMs - a.runTimeMs);
        }

        const columnPadding = [20, 6, 0];
        const pad = (s, i) => s.toString().padEnd(columnPadding[i]);

        return [
            ["Run time (ms)", "Idx", "Command"],
            ["-------------", "---", "-------"],
            ...info.map(c => [
                c.runTimeMs.toFixed(16),
                c.commandIndex,
                this._displayList.getCommandName(c)
            ])
        ].map(line => line.map(pad).join(" ")).join("\n");
    }

    write(outputDir, sort=true) {
        const fileNameBase = path.join(outputDir, this.name);
        fs.writeFileSync(`${fileNameBase}.txt`, this.formatStats(sort));

        const heatmap = new Heatmap();
        for (const info of this._commandInfo) {
            heatmap.update(info.diffPixels, info.runTimeMs);
        }

        if (!heatmap.isEmpty()) {
            heatmap.writeBitmap(`${fileNameBase}.bmp`);
        }
    }
}

// Main
function printHelp(options) {
    const valueName = (name, arg) => arg.type === "string" ? ` ${name.replace(/-/g, "_").toUpperCase()}` : "";
    const usageLine = [];
    const argLines = [];
    let argLinePad = 0;

    for (const [name, arg] of Object.entries(options)) {
        let argUsage = `-${arg.short}${valueName(name, arg)}`;
        if (!arg.required) {
            argUsage = `[${argUsage.at(-1)}]`;
        }
        usageLine.push(argUsage);

        argLines.push([`  -${arg.short}, -${name}${valueName(name, arg)}`, arg.description]);
        argLinePad = Math.max(argLinePad, argLines.at(-1)[0].length);
    }
    argLinePad += 4;

    console.log("Generates frame performance data from RSP profiler output");
    console.log(`Usage: ${path.basename(process.argv[1])} ${usageLine.join(" ")}`);
    console.log();
    console.log(`Arguments:\n${argLines.map(a => `${a[0].padEnd(argLinePad)}${a[1]}`).join("\n")}`);
}

function parseArgs(options) {
    const { values } = util.parseArgs({ options, allowPositionals: false });

    if (values.help || Object.entries(options).some(([k, v]) => !values[k] && v.required)) {
        printHelp(options);
        process.exit(1);
    }

    const camelCase = (s) => s.replace(/-(\w)/g, (_, c) => c.toUpperCase());
    return Object.fromEntries(Object.entries(values).map(([k, v]) => [camelCase(k), v]));
}

const options = {
    "debug-log-file": {
        short: "d", type: "string", required: true,
        description: "File containing debug console output"
    },
    "symbol-map-file": {
        short: "m", type: "string",
        description: "Linker-generated file containing symbol names and addresses"
    },
    "screenshot-dir": {
        short: "s", type: "string", default: process.cwd(),
        description: "Directory containing screenshot files (defaults to current directory)"
    },
    "output-dir": {
        short: "o", type: "string", default: process.cwd(),
        description: "Directory to output results to (defaults to current directory)"
    },
    "unsorted": {
        short: "u", type: "boolean",
        description: "Output the summary without sorting by command run time"
    },
    "help": {
        short: "h", type: "boolean",
        description: "Print this help text and exit"
    }
};

const { symbolMapFile, debugLogFile, screenshotDir, outputDir, unsorted } = parseArgs(options);

const symbolMap = new SymbolMap(symbolMapFile);
let currentProfile = null;

const debugLogLines = fs.readFileSync(debugLogFile, "utf-8").split('\n');
for (let i = 0; i < debugLogLines.length; ++i) {
    const line = debugLogLines[i].trim();

    if (Profile.isProfileStartLine(line)) {
        currentProfile = new Profile(symbolMap, screenshotDir);
    }

    if (currentProfile) {
        if (currentProfile.tryUpdate(line)) {
            continue;
        } else if (currentProfile.isFinished) {
            currentProfile.write(outputDir, sort=!unsorted);
            console.log(`Wrote ${currentProfile.name} to ${outputDir}`);

            currentProfile = null;
            continue;
        }
    }

    symbolMap.tryUpdate(line);
}
